### ISSUE TITLE
Investigate authorization policy blocking prometheus scraping metrics at port 15090

### ISSUE THREAD

**Issue by yangminzhu** [istio/istio#19975]
**Bug description**
Copied from a customer (Chad Wilson from ThoughtWorks ) report on slack :

> We just moved to the new AuthorizationPolicy from the old ClusterRbacConfig/ServiceRole etc. We have MTLS enforced everywhere and a deny-all type of policy for both.
One weird thing that we have found is that under the new policy Prometheus scrapes of our pods on a non-service port (configured by `prometheus.io` anotations) and scrapes of the Envoy metrics port 15090 are now blocked by the AuthorizationPolicy where they were not before.
However we don't quite understand how this happened (why it even worked before under the v1alpha1 API, or what the recommended approach is. Anyone got any pointers? 
We worked around it for now with 
```
  - to:
    - operation:
        ports:
        - "9080"
        - "15090"
```
> ...but not entirely desirable since it basically allows all access from anywhere to the ports. Trying to restrict it by service identity, paths etc doesn't seem to work; i guess because the source workload (Prometheus) is non-Mesh?

> Thanks for the clarifications (and all the great work on the new authz - feedback from myself and my team is that the new policy is easier to follow and work with, which is great. 
The change in behaviour makes sense with the switch from service-based to workload as we do not declare our monitoring ports in services, only the main app traffic port. Am I correct in inferring this puts it out-of-step with MTLS config? We enforce mTLS everywhere else in our cluster, but the proxy isn't enforcing this on the non-service monitoring ports (which is ok for us, but now seems a bit inconsistent with authZ). As far as I'm aware we're not mounting required Istio certs into Prometheus, and aren't using the prom scrape config for 'secure' pods currently.
Understand that restricting the traffic based on source namespace or principal may not be possible without an envoy around Prometheus itself.
However, is there a way to restrict the target by HTTP level attributes? Only port seems to work, i.e it seems the workload proxy doesn't know the traffic is HTTP so it blocks traffic if I add paths or method restrictions.
I'm pretty sure 15090 was also being blocked because all of the 'envoy-stats' scrapes were failing until I added the port as allowed. Ill try and export a proxy-config to see what's going on a bit later.

> I believe the only 'special' setup we have is deploying Prometheus with the stable operator chart and our own scrape configs in the operator values file into its own namespace, rather than with the Istio Helm chart.

**Comment by yangminzhu** [comment]
The user reports that Istio authorization policy is blocking Prometheus from accessing at port 15090 and this was resolved after adding the port to the authorization policy. It's a little bit strange here because the port 15090 should not be enforced by authorization policy because the Envoy listener for port 15090 is added statically.

I couldn't reproduce this for now but need more time to investigate if I missed anything here.

cc @howardjohn @douglas-reid

**Comment by chadlwilson** [comment]
I can't see anything RBAC related inside the listener config for `15090` (the below is when it's in a "broken" state, i.e Prometheus is getting `Get http://172.23.8.216:15090/stats/prometheus: EOF` errors

**Relevant config when traffic is blocked**
```json
  {
   "@type": "type.googleapis.com/envoy.admin.v2alpha.BootstrapConfigDump",
   "bootstrap": {
    "node": {
     "id": "sidecar~172.23.8.216~my-service-6484cd6dd8-5qspb.services~services.svc.cluster.local",
     "cluster": "my-service.services",
     "metadata": {
      "ISTIO_PROXY_SHA": "istio-proxy:5f5d621c9fd0b68e12c8e1b95db849588fbdc475",
      "NAMESPACE": "services",
      "INCLUDE_INBOUND_PORTS": "8080",
      "EXCHANGE_KEYS": "NAME,NAMESPACE,INSTANCE_IPS,LABELS,OWNER,PLATFORM_METADATA,WORKLOAD_NAME,CANONICAL_TELEMETRY_SERVICE,MESH_ID,SERVICE_ACCOUNT",
      "INSTANCE_IPS": "172.23.8.216",
      "pod-template-hash": "6484cd6dd8",
      "POD_PORTS": "[{\"name\":\"http\",\"containerPort\":8080,\"protocol\":\"TCP\"}]",
      "SERVICE_ACCOUNT": "my-service",
      "OWNER": "kubernetes://apis/apps/v1/namespaces/services/deployments/my-service",
      "ISTIO_VERSION": "1.4.3",
      "TRUSTJWT": "1",
      "prometheus.io/port": "9080",
      "CLUSTER_ID": "Kubernetes",
      "LABELS": {
       "app.kubernetes.io/instance": "my-service",
       "app.kubernetes.io/name": "my-service",
       "pod-template-hash": "6484cd6dd8"
      },
      "NAME": "my-service-6484cd6dd8-5qspb",
      "app.kubernetes.io/name": "my-service",
      "kubectl.kubernetes.io/restartedAt": "2020-01-13T03:24:40Z",
      "INTERCEPTION_MODE": "REDIRECT",
      "SDS": "1",
      "CONFIG_NAMESPACE": "services",
      "WORKLOAD_NAME": "my-service",
      "checksum/config": "279ea14cda38fb8cfa76f70400a76ac032222b0bccd91f8443c83150deabfa1d",
      "app.kubernetes.io/instance": "my-service",
      "kubernetes.io/psp": "eks.privileged",
      "prometheus.io/scrape": "true",
      "prometheus.io/path": "/actuator/prometheus",
      "POD_NAME": "my-service-6484cd6dd8-5qspb"
     },
     "locality": {},
     "build_version": "5f5d621c9fd0b68e12c8e1b95db849588fbdc475/1.12.0/Clean/RELEASE/BoringSSL"
    },
    "static_resources": {
     "listeners": [
      {
       "address": {
        "socket_address": {
         "address": "0.0.0.0",
         "port_value": 15090
        }
       },
       "filter_chains": [
        {
         "filters": [
          {
           "name": "envoy.http_connection_manager",
           "config": {
            "http_filters": {
             "name": "envoy.router"
            },
            "stat_prefix": "stats",
            "route_config": {
             "virtual_hosts": [
              {
               "name": "backend",
               "routes": [
                {
                 "route": {
                  "cluster": "prometheus_stats"
                 },
                 "match": {
                  "prefix": "/stats/prometheus"
                 }
                }
               ],
               "domains": [
                "*"
               ]
              }
             ]
            },
            "codec_type": "AUTO"
           }
          }
         ]
        }
       ]
      }
     ],
     "clusters": [
      {
       "name": "prometheus_stats",
       "type": "STATIC",
       "connect_timeout": "0.250s",
       "hosts": [
        {
         "socket_address": {
          "address": "127.0.0.1",
          "port_value": 15000
         }
        }
       ]
      },
      {
       "name": "xds-grpc",
       "type": "STRICT_DNS",
       "connect_timeout": "10s",
       "hosts": [
        {
         "socket_address": {
          "address": "istio-pilot.istio-system",
          "port_value": 15011
         }
        }
       ],
       "circuit_breakers": {
        "thresholds": [
         {
          "max_connections": 100000,
          "max_pending_requests": 100000,
          "max_requests": 100000
         },
         {
          "priority": "HIGH",
          "max_connections": 100000,
          "max_pending_requests": 100000,
          "max_requests": 100000
         }
        ]
       },
       "tls_context": {
        "common_tls_context": {
         "alpn_protocols": [
          "h2"
         ],
         "tls_certificate_sds_secret_configs": [
          {
           "name": "default",
           "sds_config": {
            "api_config_source": {
             "api_type": "GRPC",
             "grpc_services": [
              {
               "google_grpc": {
                "target_uri": "unix:/var/run/sds/uds_path",
                "channel_credentials": {
                 "local_credentials": {}
                },
                "call_credentials": [
                 {
                  "from_plugin": {
                   "name": "envoy.grpc_credentials.file_based_metadata",
                   "config": {
                    "header_key": "istio_sds_credentials_header-bin",
                    "secret_data": {
                     "filename": "/var/run/secrets/tokens/istio-token"
                    }
                   }
                  }
                 }
                ],
                "stat_prefix": "sdsstat",
                "credentials_factory_name": "envoy.grpc_credentials.file_based_metadata"
               }
              }
             ]
            }
           }
          }
         ],
         "combined_validation_context": {
          "default_validation_context": {
           "verify_subject_alt_name": [
            "spiffe://cluster.local/ns/istio-system/sa/istio-pilot-service-account"
           ]
          },
          "validation_context_sds_secret_config": {
           "name": "ROOTCA",
           "sds_config": {
            "api_config_source": {
             "api_type": "GRPC",
             "grpc_services": [
              {
               "google_grpc": {
                "target_uri": "unix:/var/run/sds/uds_path",
                "channel_credentials": {
                 "local_credentials": {}
                },
                "call_credentials": [
                 {
                  "from_plugin": {
                   "name": "envoy.grpc_credentials.file_based_metadata",
                   "config": {
                    "secret_data": {
                     "filename": "/var/run/secrets/tokens/istio-token"
                    },
                    "header_key": "istio_sds_credentials_header-bin"
                   }
                  }
                 }
                ],
                "stat_prefix": "sdsstat",
                "credentials_factory_name": "envoy.grpc_credentials.file_based_metadata"
               }
              }
             ]
            }
           }
          }
         }
        }
       },
       "http2_protocol_options": {},
       "dns_refresh_rate": "300s",
       "dns_lookup_family": "V4_ONLY",
       "upstream_connection_options": {
        "tcp_keepalive": {
         "keepalive_time": 300
        }
       }
      },
      {
       "name": "zipkin",
       "type": "STRICT_DNS",
       "connect_timeout": "1s",
       "hosts": [
        {
         "socket_address": {
          "address": "zipkin.istio-system",
          "port_value": 9411
         }
        }
       ],
       "dns_refresh_rate": "300s",
       "dns_lookup_family": "V4_ONLY"
      }
     ]
    },
    "dynamic_resources": {
     "lds_config": {
      "ads": {}
     },
     "cds_config": {
      "ads": {}
     },
     "ads_config": {
      "api_type": "GRPC",
      "grpc_services": [
       {
        "envoy_grpc": {
         "cluster_name": "xds-grpc"
        }
       }
      ]
     }
    },
    "tracing": {
     "http": {
      "name": "envoy.zipkin",
      "config": {
       "collector_endpoint": "/api/v1/spans",
       "collector_cluster": "zipkin",
       "trace_id_128bit": "true",
       "shared_span_context": "false"
      }
     }
    },
    "admin": {
     "access_log_path": "/dev/null",
     "address": {
      "socket_address": {
       "address": "127.0.0.1",
       "port_value": 15000
      }
     }
    },
    "stats_config": {
     "stats_tags": [
      {
       "tag_name": "cluster_name",
       "regex": "^cluster\\.((.+?(\\..+?\\.svc\\.cluster\\.local)?)\\.)"
      },
      {
       "tag_name": "tcp_prefix",
       "regex": "^tcp\\.((.*?)\\.)\\w+?$"
      },
      {
       "tag_name": "response_code",
       "regex": "(response_code=\\.=(.+?);\\.;)|_rq(_(\\.d{3}))$"
      },
      {
       "tag_name": "response_code_class",
       "regex": "_rq(_(\\dxx))$"
      },
      {
       "tag_name": "http_conn_manager_listener_prefix",
       "regex": "^listener(?=\\.).*?\\.http\\.(((?:[_.[:digit:]]*|[_\\[\\]aAbBcCdDeEfF[:digit:]]*))\\.)"
      },
      {
       "tag_name": "http_conn_manager_prefix",
       "regex": "^http\\.(((?:[_.[:digit:]]*|[_\\[\\]aAbBcCdDeEfF[:digit:]]*))\\.)"
      },
      {
       "tag_name": "listener_address",
       "regex": "^listener\\.(((?:[_.[:digit:]]*|[_\\[\\]aAbBcCdDeEfF[:digit:]]*))\\.)"
      },
      {
       "tag_name": "mongo_prefix",
       "regex": "^mongo\\.(.+?)\\.(collection|cmd|cx_|op_|delays_|decoding_)(.*?)$"
      },
      {
       "tag_name": "reporter",
       "regex": "(reporter=\\.=(.+?);\\.;)"
      },
      {
       "tag_name": "source_namespace",
       "regex": "(source_namespace=\\.=(.+?);\\.;)"
      },
      {
       "tag_name": "source_workload",
       "regex": "(source_workload=\\.=(.+?);\\.;)"
      },
      {
       "tag_name": "source_workload_namespace",
       "regex": "(source_workload_namespace=\\.=(.+?);\\.;)"
      },
      {
       "tag_name": "source_principal",
       "regex": "(source_principal=\\.=(.+?);\\.;)"
      },
      {
       "tag_name": "source_app",
       "regex": "(source_app=\\.=(.+?);\\.;)"
      },
      {
       "tag_name": "source_version",
       "regex": "(source_version=\\.=(.+?);\\.;)"
      },
      {
       "tag_name": "destination_namespace",
       "regex": "(destination_namespace=\\.=(.+?);\\.;)"
      },
      {
       "tag_name": "destination_workload",
       "regex": "(destination_workload=\\.=(.+?);\\.;)"
      },
      {
       "tag_name": "destination_workload_namespace",
       "regex": "(destination_workload_namespace=\\.=(.+?);\\.;)"
      },
      {
       "tag_name": "destination_principal",
       "regex": "(destination_principal=\\.=(.+?);\\.;)"
      },
      {
       "tag_name": "destination_app",
       "regex": "(destination_app=\\.=(.+?);\\.;)"
      },
      {
       "tag_name": "destination_version",
       "regex": "(destination_version=\\.=(.+?);\\.;)"
      },
      {
       "tag_name": "destination_service",
       "regex": "(destination_service=\\.=(.+?);\\.;)"
      },
      {
       "tag_name": "destination_service_name",
       "regex": "(destination_service_name=\\.=(.+?);\\.;)"
      },
      {
       "tag_name": "destination_service_namespace",
       "regex": "(destination_service_namespace=\\.=(.+?);\\.;)"
      },
      {
       "tag_name": "request_protocol",
       "regex": "(request_protocol=\\.=(.+?);\\.;)"
      },
      {
       "tag_name": "response_flags",
       "regex": "(response_flags=\\.=(.+?);\\.;)"
      },
      {
       "tag_name": "connection_security_policy",
       "regex": "(connection_security_policy=\\.=(.+?);\\.;)"
      },
      {
       "tag_name": "permissive_response_code",
       "regex": "(permissive_response_code=\\.=(.+?);\\.;)"
      },
      {
       "tag_name": "permissive_response_policyid",
       "regex": "(permissive_response_policyid=\\.=(.+?);\\.;)"
      },
      {
       "tag_name": "cache",
       "regex": "(cache\\.(.+?)\\.)"
      },
      {
       "tag_name": "component",
       "regex": "(component\\.(.+?)\\.)"
      },
      {
       "tag_name": "tag",
       "regex": "(tag\\.(.+?)\\.)"
      }
     ],
     "use_all_default_tags": false,
     "stats_matcher": {
      "inclusion_list": {
       "patterns": [
        {
         "prefix": "reporter="
        },
        {
         "prefix": "component"
        },
        {
         "prefix": "cluster_manager"
        },
        {
         "prefix": "listener_manager"
        },
        {
         "prefix": "http_mixer_filter"
        },
        {
         "prefix": "tcp_mixer_filter"
        },
        {
         "prefix": "server"
        },
        {
         "prefix": "cluster.xds-grpc"
        },
        {
         "suffix": "ssl_context_update_by_sds"
        }
       ]
      }
     }
    }
   },
   "last_updated": "2020-01-13T08:42:38.424Z"
  },
```

```json
{
   "@type": "type.googleapis.com/envoy.admin.v2alpha.ListenersConfigDump",
   "version_info": "2020-01-14T04:05:23Z/28",
   "static_listeners": [
    {
     "listener": {
      "address": {
       "socket_address": {
        "address": "0.0.0.0",
        "port_value": 15090
       }
      },
      "filter_chains": [
       {
        "filters": [
         {
          "name": "envoy.http_connection_manager",
          "config": {
           "route_config": {
            "virtual_hosts": [
             {
              "name": "backend",
              "routes": [
               {
                "match": {
                 "prefix": "/stats/prometheus"
                },
                "route": {
                 "cluster": "prometheus_stats"
                }
               }
              ],
              "domains": [
               "*"
              ]
             }
            ]
           },
           "codec_type": "AUTO",
           "http_filters": {
            "name": "envoy.router"
           },
           "stat_prefix": "stats"
          }
         }
        ]
       }
      ]
     },
     "last_updated": "2020-01-13T08:42:38.502Z"
    }
   ],
   "dynamic_active_listeners":
  # ...
```

There are no other references to the port 15090 listener in the config dump.

When I allow `15090` in the `AuthorizationPolicy` the only differences in the config are

```json
                     {
                      "destination_port": 15090
                     }
```
^^^ is added to the `type.googleapis.com/envoy.config.filter.network.rbac.v2.RBAC` config `or_rules` for 
* port `8080` listener (main service traffic port), as expected
* port `15006` `virtualInbound` TCP proxy listener

Are there any other relevant bits? It's a bit difficult for me to safely redact stuff from the full config dump.

**Comment by moritzheiber** [comment]
I can confirm that this is an issue, for us the only remedy was:

```
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: deny-all
  namespace: my-namespace
spec:
  {}

---
---
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: my-app
  namespace: my-namespace
spec:
  selector:
    matchLabels:
      app: my-app
  rules:
  - from:
    - source:
        namespaces: ["prometheus-namespace"]
    to:
    - operation:
        methods: ["GET"]
        paths: ["/metrics"]
  - to:
    - operation:
        ports: ["15090"]
```

Notice the difference for the app-specific Prometheus rule (works with `source` and a `namespaces` target in its `operation`) and the Envoy proxy specific rules (which **must** only carry a single `to` rule with a `port` in `operation`, otherwise it's discarded).

This is a sample request from the operator:

```
{"bytes_sent":"31863","upstream_cluster":"InboundPassthroughClusterIpv4","downstream_remote_address":"10.0.129.14:43418","authority":"-","path":"-","protocol":"-","upstream_service_time":"-","upstream_local_ │
│ address":"127.0.0.6:55781","duration":"21","downstream_local_address":"10.0.152.1:15090","upstream_transport_failure_reason":"-","route_name":"-","response_code":"0","user_agent":"-","response_flags":"-","st │
│ art_time":"2020-02-04T14:21:56.091Z","method":"-","request_id":"-","upstream_host":"10.0.152.1:15090","x_forwarded_for":"-","requested_server_name":"-","bytes_received":"95","istio_policy_status":"-"}
```

The Prometheus operator is running _outside_ of the Istio mesh in this case.

**Comment by chadlwilson** [comment]
@yangminzhu did we ever manage to look at this? This is still a problem on `1.4.7`. Haven't yet managed to upgrade to `1.5.1` successfully to re-validate.

**Comment by yangminzhu** [comment]
> When I allow 15090 in the AuthorizationPolicy the only differences in the config are
>
>                     {
>                      "destination_port": 15090
>                     }
>^^^ is added to the type.googleapis.com/envoy.config.filter.network.rbac.v2.RBAC config or_rules for
>
>port 8080 listener (main service traffic port), as expected
>port 15006 virtualInbound TCP proxy listener
> Are there any other relevant bits? It's a bit difficult for me to safely redact stuff from the full config dump.

Sorry I forgot to reply to this, I think this clearly means the 15090 port is handled by the virtual inbound instead of by the static configured listener, that's why it's being rejected by AuthorizationPolicy because the AuthorizationPolicy applies to the virtual inbound listener (but not static listener). This is actually working as expected from the AuthorizationPolicy perspective.

It sounds like a bug that the 15090 is not handled by the static listener, @lambdai @douglas-reid could you help take a look and confirm is it a bug that the 15090 not handled by the static listener but instead by the virtual inbound listener? thank you.

For now, I think it's fine to create an authorization policy to whitelist the 15090 port. This shouldn't cause security issues because if the 15090 is handled correctly by the static listener, it won't have any access control anyway. However this is inconvenience and more of a workaround for the bug that 15090 not handled by the static listener.

**Comment by chadlwilson** [comment]
OK, noted, thanks for the investigation and reply as always @yangminzhu :-)

**Comment by lambdai** [comment]
We have a specialized setting for port 15090
More precisely

1. The iptables doesn't capture traffic to 15090.
2. 15090 is a reserved port in istio's proxy which has static config: including listener, route, cluster.

With the **static** config, there is no customized setting to add authorization on port 15090. 
That says, istio only support plain text access to port 15090. Any attempt to add mtls on 15090 is ignored at the envoy listening on 15090.

If you enable mesh side mtls, I think the policy somehow affect the sidecar proxy of the client which initiate the request to 15090, that's why @chadlwilson is seeing request failure.

**Comment by lambdai** [comment]
Two possible solutions

1. authz ignores any policy at client side setting on 15090, aligning with the de factor that 15090 is statically configured as plaintext
2. Drive the networking setting to build an end2end solution: envoy bootstrap config, iptables, dynamic generate LDS/RDS/CDS respecting security policy.

**Comment by chadlwilson** [comment]
> If you enable mesh side mtls, I think the policy somehow affect the sidecar proxy of the client which initiate the request to 15090, that's why @chadlwilson is seeing request failure.

I don't think that can be the case here, because the calling service is Prometheus; and in our case it has no sidecar proxy; it's not part of the Mesh - and we haven't attempted to make our Prometheus be able to do mTLS with Mesh-proxied-services.

**Comment by lambdai** [comment]
>  and in our case it has no sidecar proxy; it's not part of the Mesh - and we haven't attempted to make our Prometheus be able to do mTLS with Mesh-proxied-services

That's a good call. So there is only one possibility: the iptables capture.

@chadlwilson 
1. Could you paste your iptables rule?  
2. Are you using istio-cni? I realize 15090 is not in the exclude list if you use istio-cni and leading to your case. [Fixed days ago](https://github.com/istio/cni/pull/299)

**Comment by yangminzhu** [comment]
> > and in our case it has no sidecar proxy; it's not part of the Mesh - and we haven't attempted to make our Prometheus be able to do mTLS with Mesh-proxied-services
> 
> That's a good call. So there is only one possibility: the iptables capture.
> 
> @chadlwilson
> 
> 1. Could you paste your iptables rule?
> 2. Are you using istio-cni? I realize 15090 is not in the exclude list if you use istio-cni and leading to your case. [Fixed days ago](https://github.com/istio/cni/pull/299)

I can reproduce this with a fresh install of 1.4.7, the following is the iptable rules for httpbin:
```
$ k logs -f httpbin-654c6cbbb9-48rqv -c istio-init 
Environment:
------------
ENVOY_PORT=
INBOUND_CAPTURE_PORT=
ISTIO_INBOUND_INTERCEPTION_MODE=
ISTIO_INBOUND_TPROXY_MARK=
ISTIO_INBOUND_TPROXY_ROUTE_TABLE=
ISTIO_INBOUND_PORTS=
ISTIO_LOCAL_EXCLUDE_PORTS=
ISTIO_SERVICE_CIDR=
ISTIO_SERVICE_EXCLUDE_CIDR=

Variables:
----------
PROXY_PORT=15001
PROXY_INBOUND_CAPTURE_PORT=15006
PROXY_UID=1337
PROXY_GID=1337
INBOUND_INTERCEPTION_MODE=REDIRECT
INBOUND_TPROXY_MARK=1337
INBOUND_TPROXY_ROUTE_TABLE=133
INBOUND_PORTS_INCLUDE=*
INBOUND_PORTS_EXCLUDE=15020
OUTBOUND_IP_RANGES_INCLUDE=*
OUTBOUND_IP_RANGES_EXCLUDE=
OUTBOUND_PORTS_EXCLUDE=
KUBEVIRT_INTERFACES=
ENABLE_INBOUND_IPV6=

+ iptables -t nat -N ISTIO_REDIRECT
+ iptables -t nat -A ISTIO_REDIRECT -p tcp -j REDIRECT --to-port 15001
+ iptables -t nat -N ISTIO_IN_REDIRECT
+ '[' '*' == '*' ']'
+ iptables -t nat -A ISTIO_IN_REDIRECT -p tcp -j REDIRECT --to-port 15006
+ '[' -n '*' ']'
+ '[' REDIRECT = TPROXY ']'
+ table=nat
+ iptables -t nat -N ISTIO_INBOUND
+ iptables -t nat -A PREROUTING -p tcp -j ISTIO_INBOUND
+ '[' '*' == '*' ']'
+ iptables -t nat -A ISTIO_INBOUND -p tcp --dport 22 -j RETURN
+ '[' -n 15020 ']'
+ for port in ${INBOUND_PORTS_EXCLUDE}
+ iptables -t nat -A ISTIO_INBOUND -p tcp --dport 15020 -j RETURN
+ '[' REDIRECT = TPROXY ']'
+ iptables -t nat -A ISTIO_INBOUND -p tcp -j ISTIO_IN_REDIRECT
+ iptables -t nat -N ISTIO_OUTPUT
+ iptables -t nat -A OUTPUT -p tcp -j ISTIO_OUTPUT
+ '[' -n '' ']'
+ iptables -t nat -A ISTIO_OUTPUT -o lo -s 127.0.0.6/32 -j RETURN
+ '[' -z '' ']'
+ iptables -t nat -A ISTIO_OUTPUT -o lo '!' -d 127.0.0.1/32 -j ISTIO_IN_REDIRECT
+ for uid in ${PROXY_UID}
+ iptables -t nat -A ISTIO_OUTPUT -m owner --uid-owner 1337 -j RETURN
+ for gid in ${PROXY_GID}
+ iptables -t nat -A ISTIO_OUTPUT -m owner --gid-owner 1337 -j RETURN
+ iptables -t nat -A ISTIO_OUTPUT -d 127.0.0.1/32 -j RETURN
+ '[' 0 -gt 0 ']'
+ '[' 1 -gt 0 ']'
+ '[' '*' == '*' ']'
+ iptables -t nat -A ISTIO_OUTPUT -j ISTIO_REDIRECT
+ set +o nounset
+ '[' -n '' ']'
+ ip6tables -F INPUT
+ ip6tables -A INPUT -m state --state ESTABLISHED -j ACCEPT
+ ip6tables -A INPUT -i lo -d ::1 -j ACCEPT
+ ip6tables -A INPUT -j REJECT
+ dump
+ iptables-save
# Generated by iptables-save v1.6.1 on Thu Apr 23 01:06:14 2020
*nat
:PREROUTING ACCEPT [0:0]
:INPUT ACCEPT [0:0]
:OUTPUT ACCEPT [0:0]
:POSTROUTING ACCEPT [0:0]
:ISTIO_INBOUND - [0:0]
:ISTIO_IN_REDIRECT - [0:0]
:ISTIO_OUTPUT - [0:0]
:ISTIO_REDIRECT - [0:0]
-A PREROUTING -p tcp -j ISTIO_INBOUND
-A OUTPUT -p tcp -j ISTIO_OUTPUT
-A ISTIO_INBOUND -p tcp -m tcp --dport 22 -j RETURN
-A ISTIO_INBOUND -p tcp -m tcp --dport 15020 -j RETURN
-A ISTIO_INBOUND -p tcp -j ISTIO_IN_REDIRECT
-A ISTIO_IN_REDIRECT -p tcp -j REDIRECT --to-ports 15006
-A ISTIO_OUTPUT -s 127.0.0.6/32 -o lo -j RETURN
-A ISTIO_OUTPUT ! -d 127.0.0.1/32 -o lo -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -m owner --uid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -m owner --gid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -d 127.0.0.1/32 -j RETURN
-A ISTIO_OUTPUT -j ISTIO_REDIRECT
-A ISTIO_REDIRECT -p tcp -j REDIRECT --to-ports 15001
COMMIT
# Completed on Thu Apr 23 01:06:14 2020
+ ip6tables-save
# Generated by ip6tables-save v1.6.1 on Thu Apr 23 01:06:14 2020
*filter
:INPUT ACCEPT [0:0]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [0:0]
-A INPUT -m state --state ESTABLISHED -j ACCEPT
-A INPUT -d ::1/128 -i lo -j ACCEPT
-A INPUT -j REJECT --reject-with icmp6-port-unreachable
COMMIT
# Completed on Thu Apr 23 01:06:14 2020
```

**Comment by lambdai** [comment]
```
INBOUND_PORTS_EXCLUDE=15090,15020

-A ISTIO_INBOUND -p tcp --dport 22 -j RETURN
-A ISTIO_INBOUND -p tcp --dport 15090 -j RETURN
-A ISTIO_INBOUND -p tcp --dport 15020 -j RETURN
```
Interesting... above is mine on 1.5.1

The workaround is to set INBOUND_PORTS_EXCLUDE by adding 15090 to deployment annotation [traffic.sidecar.istio.io/excludeInboundPorts](https://istio.io/docs/reference/config/annotations/)
This workaround is harmless when you upgrade to istio 1.5

**Comment by koalaty-code** [comment]
The `envoy-stats` job is failing to scrape any targets that have an AuthorizationPolicy defined on Istio 1.5 and 1.6. Prometheus errors with `server returned HTTP status 403 Forbidden`. Aside from setting `INBOUND_PORTS_EXCLUDE`, updating the AuthorizationPolicy to allow anonymous requests on path `/stats/prometheus` fixes the issue.

Is there a work-around that can be applied globally in Istio or Prometheus configuration that doesn't require updating every workload using an AuthorizationPolicy?

**Comment by yangminzhu** [comment]
@koalaty-code if you're still using 1.5, you can apply the authorization policy in mesh-level (no workload selector and in the root namespace which is istio-system by default), it should then apply to all workloads in the mesh.

**Comment by shadjac** [comment]
I have same setup mentioned in bug description. Prometheus is not in istio-mesh (namespace= monitoring), it is in different namespace that the services(namespace= services)
We do not want some services to talk to each other. We were using whitelisting of service accounts in ClusterRbacConfig
But after moving to 1.7.3, we had to whitelist `8080` (main service container port) to fix `Get http://172.23.8.216:8080/actuator/prometheus: EOF`. As a side effect, all services can now talk to each other as Chad Wilson mentioned

> ...but not entirely desirable since it basically allows all access from anywhere to the ports. Trying to restrict it by service 
   identity, paths etc doesn't seem to work; i guess because the source workload (Prometheus) is non-Mesh?
 Is there any solution for this yet?

**Comment by chadlwilson** [comment]
@shadjac I'm not aware of any direct option to resolve this; however I am guessing that you have at least two possible options now

1. Move Spring Boot Actuator endpoints to a different port that you are comfortable exposing without RBAC or mTLS via `management.server.port=9080` allowing you to keep RBAC+mTLS on your main app port
2. Try out [Istio 1.6+'s Prometheus metrics merging](https://istio.io/latest/docs/ops/integrations/prometheus/#option-1-metrics-merging) (never tried this personally) which I believe will allow you to keep scraping with non-TLS via the proxy's status port, as long as you use `prometheus.io` annotations rather than `(Pod|Service)Monitor`s.

**Comment by shadjac** [comment]
@chadlwilson Thank you for these suggestions! We wanted to choose 2nd. But looks like for metrics merging to work, the application metrics should exist at ":15020/stats/prometheus" which brings us back to suggestion 1st- "change port in application". We could scrape without using service monitors but currently we push metrics to different port and different path than this - ":15020/stats/prometheus". I could not find more information on this in documentation.
My assumption now is - if we bring prometheus under service mesh, we will be able to use authorization policy by whitelisting traffic using Service Identity.
Also, is this still an issue in 1.8.x if anybody has tried that out?
